### PR TITLE
Tests: Updated device list; Modified BrowserStackLocal startup

### PR DESCRIPTION
**Pull request by punitx**

### CODE REVIEW

File: test/karma/karma.conf.js [hunk 1]
```diff
@@ -73,6 +73,7 @@ module.exports = function( config ) {
 		colors: !isTravis,
 
 		port: 9876,
+		hostname: "bs-local.com",
```
Comment (mgol) [inline test/karma/karma.conf.js]: Where does that hostname come from?

File: test/karma/launchers.js [hunk 2]
```diff
@@ -211,77 +211,96 @@ module.exports = {
 		os_version: "High Sierra"
 	},
 
-	"bs_ios-5.1": {
+	"bs_iphone_7-10.3": {
```
Comment (mgol) [inline test/karma/launchers.js]: Sizzle supports iOS 5.1+ so don't remove older devices. Also, we don't need two copies of the same iOS version like 10.3 here.

File: test/karma/launchers.js [hunk 3]
```diff
@@ -211,77 +211,89 @@ module.exports = {
...
-		device: "iPhone 6S",
+		device: "iPhone X",
 		os: "ios",
-		os_version: "9.3"
+		os_version: "11.0",
+		real_mobile: true
 	},
-	"bs_ios-10.3": {
+
+	"bs_android-4.4": {
```
Comment (mgol) [inline test/karma/launchers.js]: Sizzle supports Android 2.3+, unfortunately...

**Comment by mgol** [comment]
Note that our PRs run only on PhantomJS for security reasons. To test how it really works you'd have to set up Travis on your account and add a special commit on a branch outside of this PR that will switch the BrowserStack credentials to yours; then you'll see a full run.

File: test/karma/karma.conf.js (hunk 1 above)
Comment (punitx) [reply test/karma/karma.conf.js]: It is needed to run tests on physical iOS devices as those devices don't support forwarding request for loopback addresses to the proxy. Also, `bs-local.com` points to 127.0.0.1. So it shouldn't create an issue while running tests anywhere.

Comment (mgol) [reply test/karma/karma.conf.js]: Nice, I didn't know about this! Could you add a comment above this line to explain it?

Comment (mgol) [reply test/karma/karma.conf.js]: One problem with it - this hostname won't work when tests are run locally, will it? Perhaps we should set it to `"bs-local.com"` only on Travis and only when BrowserStack is used (i.e. not in PRs)? I'm actually surprised PhantomJS test runs succeeded on Travis with this setting enabled; do you know why?

**Comment by punitx** [comment]
I really appreciate the heads up regarding testing on Travis. I have run a few [builds](https://travis-ci.org/punitx/sizzle/builds) on Travis. Current PR is passing. I will make some additional changes to include older versions of Android and iOS devices as well.

File: Gruntfile.js [hunk 4]
```diff
@@ -271,10 +271,10 @@ module.exports = function( grunt ) {
 		"karma:ios",
 
 		"karma:oldIe", "karma:oldFirefox", "karma:oldChrome",
-		"karma:oldSafari", "karma:oldOpera"
+		"karma:oldSafari", "karma:oldOpera",
 
 		// See #314 :-(
-		// "karma:android", "karma:oldAndroid"
+		"karma:android", "karma:oldAndroid"
```
Comment (mgol) [inline Gruntfile.js]: The comment above this line no longer applies.

File: Gruntfile.js [hunk 5]
```diff
@@ -62,7 +62,7 @@ module.exports = function( grunt ) {
 			safari: [ "bs_safari-4.0", "bs_safari-5.0", "bs_safari-5.1" ],
 			ie: [ "bs_ie-7", "bs_ie-8" ],
 			opera: [ "bs_opera-11.6", "bs_opera-12.16" ],
-			android: [ "bs_android-2.3" ]
+			android: [ "bs_android-4.4" ]
```
Comment (mgol) [inline Gruntfile.js]: Could you add a comment above this line that older Android devices are too unreliable to be included in CI (and finish that with `see #314` as was done below)?

File: test/karma/karma.conf.js [hunk 6]
```diff
@@ -86,7 +87,8 @@ module.exports = function( config ) {
 		captureTimeout: 3e5,
 		browserNoActivityTimeout: 3e5,
 		browserDisconnectTimeout: 3e5,
-		browserDisconnectTolerance: 3
+		browserDisconnectTolerance: 3,
+		startTunnel: false
```
Comment (mgol) [inline test/karma/karma.conf.js]: Please add an empty line above this one and a line with a comment saying the tunnel is started manually in `.travis.yml` for stability reasons.

File: test/karma/launchers.js (hunk 2 above)
Comment (mgol) [reply test/karma/launchers.js]: OK, so we can't really add old Android versions back as they're too unstable (and will be removed in a few months) but please only leave one entry for each iOS version and name them like they were named currently, i.e. `bs_ios_VERSION`.

File: test/karma/launchers.js [hunk 7]
```diff
@@ -211,77 +211,75 @@ module.exports = {
...
 		base: "BrowserStack",
-		device: "iPhone 5S",
+		device: "iPhone 7",
 		os: "ios",
-		os_version: "7.0"
+		os_version: "10.3",
+		real_mobile: true
 	},
-	"bs_ios-8.3": {
+	"bs_ios-11.0": {
```
Comment (mgol) [inline test/karma/launchers.js]: I think we're fine with the latest version of each iOS major, especially that the API usually deletes the older ones.

File: Gruntfile.js [hunk 8]
```diff
@@ -48,12 +48,11 @@ module.exports = function( grunt ) {
 		];
 
 		browsers.ios = [
-			"bs_ios-5.1", "bs_ios-6.0", "bs_ios-7.0", "bs_ios-8.3",
-			"bs_ios-9.3", "bs_ios-10.3"
+			"bs_ios-10.3", "bs_ios-11.0", "bs_ios-11.2"
```
Comment (mgol) [inline Gruntfile.js]: This removes a lot of iOS versions and iOS simulators seem to be pretty stable compared to Android ones. Could you leave iOS as it was, only adding 11.2?

**Comment by mgol** [comment]
Thank you. I won’t be available for the next few weeks to do a final review
& land this; I’ll have a look afterwards.
-- 
Michał Gołębiowski-Owczarek

**Comment by mgol** [comment]
Sorry for letting it languish. With #434 we were able to add new iOS devices while using the `bs-local.com` domain. I tried to rebase this PR over current master both with and without using the manually created tunnels but in both cases the tests were behaving weirdly with some browsers being spawned twice etc. so it wasn't working.

I'll close this PR then and will ask BrowserStack separately if there's anything we can do now to test in Android devices.

Thank you for your efforts.

**Pull request closed without merging**